Changed from static state/transition creation to dynamic

davegiannini opened the pull request:
See [Go State Machine Framework](https://confluence.enflick.com/x/9AIgAw) for the object model associated with these code changes.  Here's what changed:

1. Updated framework to match above design
1. States and transitions are dynamically created at run-time so that external conditions can be evaluated when deciding on a next state
1. Re-wrote samek example to use updated framework
1. Re-wrote simple example to use updated framework

davegiannini committed:
Changed from static state/transition creation to dynamic

chrisleavoy commented on transition.go:
```diff
@@ -1,49 +1,89 @@
-// EndTransition is a convenience variable representing a transition to an end state.
-var EndTransition = &DirectTransition{
-	next: EndState,
+type DaveEndTransition struct {
```
what is a "Dave" and why is it in the library? ;)

davegiannini replied:
Fixed.  This was left over from an earlier refactor.

davegiannini committed:
Applied go best practices for getters and comments

davegiannini committed:
Added unit tests and updated simple integration test

davegiannini committed:
Removed EventHandler and StateEngine and renamed StateMachine

AlexRomanEnflick committed:
Initial custom context work

AlexRomanEnflick committed:
A few changes...

- Change Context name to Service
- And (I hope) everything else that was named context or ctx
- Also adapt samek example to the new model
- Add example of how to extend a hsm.Service for use in states

AlexRomanEnflick committed:
Fix unit tests and renamed a couple of things

AlexRomanEnflick committed:
Clean up how Service is passed around

AlexRomanEnflick committed:
Address PR feedback

AlexRomanEnflick committed:
Merge pull request #7 from Enflick/alex_context

Add the concept of a Service to the state machine framework

rmrobinson-textnow commented on state.go:
```diff
@@ -0,0 +1,54 @@
+
+// IsNilState returns true if the passed in state is of type UndefinedState
+func IsNilState(state State) bool {
+	return reflect.TypeOf(state).String() == "*hsm.UndefinedState"
```
The 'reflect' package is typically avoided due to runtime overhead.
You can accomplish the same by using a type assertionand checking the result, i.e.
```
_, ok := transition.(*UndefinedTransition)
return ok
```
See https://tour.golang.org/methods/15 for details on this language feature.

davegiannini replied:
Thanks, I'll make the change.

rmrobinson-textnow commented on mock_event.go:
```diff
@@ -0,0 +1,17 @@
+package hsm
```
We should move the mock_ files to be suffixed with _test.go to avoid them being considered part of the public API of the package. If the are in _test.go files they will only be compiled and used when go test is run.

davegiannini replied:
Thanks, I'll make the change.

rmrobinson-textnow commented on service.go:
```diff
@@ -0,0 +1,33 @@
+
+import "go.uber.org/zap"
+
+type Service interface {
```
I'm not sure if this is needed to be part of this package, as none of the pieces of this package use it. This may be used by an implementation of the hsm package, but the Service itself seems like a separate component. This style of Get and Set on interfaces is pretty non-idiomatic as you lose all sorts of type safety, so I'd avoid including it if it isn't used.

davegiannini replied:
Agreed.

rmrobinson-textnow commented on stateMachine.go:
```diff
@@ -0,0 +1,137 @@
+/*
+HSM provides the framework for Hierarchical State Machine implementations.
```
I think this is a really good writeup. It would probably be useful to move into the README, instead of here, so that new users of the package are able to quickly see how things are laid out.
Also, since this is an open source project I'd avoid including links to our internal confluence page; if we can summarize the essentials either in the README or in a dependent package that will make it easier for others to use.

davegiannini replied:
Good, I'll remove the internal links and external links and add the object model as a png.

From my reading up on go best practices I learned that I should add comments like these in one file associated with the package for godoc rendering.  I also like the README.md pattern.  For now I'll copy the content into both locations.

rmrobinson-textnow replied:
Makes sense! 👍

rmrobinson-textnow commented on stateMachine.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+// StateMachine constructor
+func NewStateMachine(svc Service, startState State) *StateMachine {
```
I would not depend on the service having a logger, and pass the logger in as an explicit dependency here. It is cleaner when setting up and logging events in the state machine.

rmrobinson-textnow replied:
Avoiding passing in a Service handle was actually a key design decision I made with the original state machine framework, as it forces a layer of knowledge about the usage of the framework into the framework itself. Looking through this I don't think we actually need to pass the Service in at all; any registered actions (which are today taking in the service) can be implemented as methods on the Service (which will prevent us needing to supply the Service to the action invocation). The Service is only used by Execute to supply to the Action so I think this change will clean this up.

davegiannini replied:
Okay.  So instead we should pass everything we need to our State constructor to get the most out of the compiler to find bugs?  For example, Logger, ESLConnection, RedisConnection, HTTPConnection, etc?

I think this makes sense.

rmrobinson-textnow replied:
Yes exactly. Each state may take a different set of parameters; so we can be sure we're properly initializing each state. It also makes testing the states easier as they aren't dependent on a properly-set-up service. Since the state machine doesn't create any states itself, there could be possible state machine implementations where the service pattern is superfluous, or cases where multiple services may be desired, so keeping it up to the state to properly initialize its children will ensure the least amount of coupling between our state machine and the states that it manages.

rmrobinson-textnow commented on stateMachine.go:
```diff
@@ -0,0 +1,137 @@
+/*
```
Small nit, but files are idiomatically named with snake case vs. camel case (ie. state_machine.go)

davegiannini replied:
Thanks, I'll make the change.

rmrobinson-textnow commented on transition.go:
```diff
@@ -1,49 +1,120 @@
+
+// IsNilTransition returns true if the passed in transition is of type UndefinedTransition
+func IsNilTransition(transition Transition) bool {
+	return reflect.TypeOf(transition).String() == "*hsm.UndefinedTransition"
```
Same comment about using a type assertion vs. reflection.

davegiannini replied:
Thanks, I'll make the change.

rmrobinson-textnow commented on precondition.go:
```diff
@@ -0,0 +1,8 @@
+package hsm
+
+func Precondition(srv Service, expression bool, message string) {
```
Let's just pass the logger in vs. the service.

davegiannini replied:
Thanks, I'll make the change.

rmrobinson-textnow reviewed (changes requested):
I didn't set up a pre-commit hook, but if you run
```
golint .
```
at the top level you'll see a number of warnings. Most are comment-related, but we want to clean these up. We should also run 'go fmt' to ensure the files are formatted appropriately.

chrisleavoy commented on transition.go:
```diff
@@ -1,49 +1,120 @@
+	//   - parentState is NilState
+	//   - parentState is equal to toState's parentState
+	parentState := fromState.OnExit(t.event)
+	for !IsNilState(parentState) {
```
This looks like an unsafe infinite loop. For prod code, should there be a guard somewhere that prevents cycles in the hsm?

davegiannini replied:
The only way this can be infinite is if there is a circular state machine implementation.  However, that can happen so good catch.  I'll add code and a test to check and handle this condition.

chrisleavoy commented on precondition.go:
```diff
@@ -0,0 +1,8 @@
+func Precondition(srv Service, expression bool, message string) {
+	if !expression {
+		srv.Logger().Panic(message)
+		panic(message)
```
Panic is an anti pattern. Especially in a library, should be removed. 

I’m also surprised not to find error built into the library at all. Is it designed to be infallible?

davegiannini replied:
I read up that panic should be used in Go to express programming errors.  Preconditions resulting in panics should be used when someone is calling a method and not obeying the method's contract.  My tendency is to let the program crash (even server code) on programming errors to make them easy to debug and quickly fix.  Otherwise, the program is likely to have undefined behaviour and harder to debug later when it eventually crashes for other reasons or exhibits resource problems.

The framework has a minimal number of methods, all designed such that they need to handle their errors and never return error since the framework's calling code will never know how to properly handle an error.

rmrobinson-textnow replied:
logger.Panic() itself panics, so the second line is redundant :)

AlexRomanEnflick committed:
Make a few other needed changes

chrisleavoy commented on example/simple/states/stateA.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (s *StateA) OnExit(event hsm.Event) hsm.State {
+	hsm.Precondition(s.srv, !s.exited, fmt.Sprintf("State %s has already been entered", s.Name()))
```
wrong error description

davegiannini replied:
Fixed

davegiannini committed:
Code review updates

AlexRomanEnflick committed:
fix import

rmrobinson-textnow commented on example/simple/states/state_a.go:
```diff
@@ -0,0 +1,93 @@
+func (s *StateA) EventHandler(event hsm.Event) hsm.Transition {
+	switch event.ID() {
+	case ec.ID():
+		return hsm.NewExternalTransition(event, NewStateA(s.logger, s.a), action3)
```
We can probably just return s here instead of creating a NewStateA on this transition?

davegiannini replied:
We need to construct a new state since, by design, Enter/Exit can only be called once per state instance.

rmrobinson-textnow commented on example/simple/states/state_b.go:
```diff
@@ -0,0 +1,75 @@
+	"go.uber.org/zap"
+)
+
+// StateA represents State A
```
Typo.

davegiannini replied:
Fixed

rmrobinson-textnow commented on example/simple/states/state_c.go:
```diff
@@ -0,0 +1,83 @@
+	"go.uber.org/zap"
+)
+
+// StateA represents State A
```
Typo

davegiannini replied:
Fixed

rmrobinson-textnow commented on example/simple/states/state_d.go:
```diff
@@ -0,0 +1,72 @@
+	"go.uber.org/zap"
+)
+
+// StateA represents State A
```
Typo

davegiannini replied:
Fixed

rmrobinson-textnow commented on assertion.go:
```diff
@@ -0,0 +1,13 @@
+// Assertion causes an runtime panic if expression is not true
+func Assertion(logger *zap.Logger, expression bool, message string) {
+	if !expression {
+		logger.Panic(message)
```
same observation about logger.Panic itself panic-ing.

davegiannini replied:
Fixed

davegiannini committed:
Code review updates
